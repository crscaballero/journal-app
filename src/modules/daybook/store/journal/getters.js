export const getEntriesByTerm = (state) => (term = '') => {
    if (!term.length) return state.entries
    return state.entries.filter(entry => entry.text.toLowerCase().includes(term.toLowerCase()));
}

export const getEntryById = (state) => (id = '') => {
    const entry = state.entries.find(entry => entry.id === id);
    if (!entry) return;
    return {...entry};
}
