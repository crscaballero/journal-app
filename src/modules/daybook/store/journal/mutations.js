export const setEntries = (state, entries) => {
    state.entries = [...state.entries, ...entries];
    state.isLoading = false;
}

export const updateEntry = (state, entry) => {
    // console.log('updateEntry:entry:', entry);
    const idx = state.entries.map(e => e.id).indexOf(entry.id);
    state.entries[idx] = entry;
}

export const addEntry = (state, entry) => {
    // console.log('addEntry:entry:', entry);
    state.entries.unshift(entry);
}

export const deleteEntry = (state, id) => {
    // console.log('deleteEntry:id:', id);
    const index = state.entries.findIndex(entry => entry.id === id);
    // console.log('index:', index);
    state.entries.splice(index, 1);
}
