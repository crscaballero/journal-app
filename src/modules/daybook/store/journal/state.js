export default () => ({
    isLoading: true,
    entries: [
        {
            id: '1',
            date: new Date().toDateString(),
            text: 'Lorem ipsum dolor sit amet consectetur adipisicing elit. Debitis, fugiat ex eum pariatur cupiditate reiciendis magnam ipsam, maxime aspernatur, architecto doloribus eos quasi molestias porro illum repellat inventore atque expedita.',
            picture: 'https://cdn.vallarta-adventures.com/sites/default/files/2021-05/CHichen10.jpeg'
        },
        {
            id: '2',
            date: new Date().toDateString(),
            text: 'Lorem ipsum dolor sit amet consectetur adipisicing elit. Debitis, fugiat ex eum pariatur cupiditate reiciendis magnam ipsam, maxime aspernatur, architecto doloribus eos quasi molestias porro illum repellat inventore atque expedita.',
            picture: null
        },
        {
            id: '3',
            date: new Date().toDateString(),
            text: 'Lorem ipsum dolor sit amet consectetur adipisicing elit. Debitis, fugiat ex eum pariatur cupiditate reiciendis magnam ipsam, maxime aspernatur, architecto doloribus eos quasi molestias porro illum repellat inventore atque expedita.',
            picture: null
        },
        {
            id: '4',
            date: new Date().toDateString(),
            text: 'Lorem ipsum dolor sit amet consectetur adipisicing elit. Debitis, fugiat ex eum pariatur cupiditate reiciendis magnam ipsam, maxime aspernatur, architecto doloribus eos quasi molestias porro illum repellat inventore atque expedita.',
            picture: null
        },
        {
            id: '5',
            date: new Date().toDateString(),
            text: 'Lorem ipsum dolor sit amet consectetur adipisicing elit. Debitis, fugiat ex eum pariatur cupiditate reiciendis magnam ipsam, maxime aspernatur, architecto doloribus eos quasi molestias porro illum repellat inventore atque expedita.',
            picture: null
        }
    ]
});
