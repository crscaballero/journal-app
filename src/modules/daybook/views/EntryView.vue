<template>
    <template v-if="entry">
        <div class="entry-title d-flex justify-content-between p-2" >
            <div>
                <span class="text-success fs-3 fw-bold">{{day}}</span>
                <span class="mx-1 fs-3">{{month}}</span>
                <span class="textmx-2 fs-4 fw-light">{{yearDay}}</span>
            </div>
            <div>
                <button
                    v-if="entry.id"
                    @click="onDeleteEntry"
                    type="button"
                    class="btn btn-danger mx-2"
                >
                    Delete <i class="fa fa-trash-all"></i>
                </button>
                <input
                    type="file"
                    @change="onSelectedImage"
                    ref="imageSelector"
                    accept="image/png, image/jpeg"
                    v-show="false"
                />
                <button
                    v-if="entry.id"
                    type="button"
                    class="btn btn-primary"
                    @click="onSelectImage"
                >
                    Upload image <i class="fa fa-upload"></i>
                </button>
            </div>
        </div>
        <hr />
        <div class="d-flex flex-column px-3 h-75">
            <textarea
                v-model="entry.text"
                cols="30"
                rows="10"
                placeholder="What happened today?..."
            ></textarea>
        </div>
        <Fab
            icon="fa-save"
            @on:click="saveEntry"
        />
        <img
            v-if="entry?.picture && !localImage"
            :src="entry.picture"
            alt="entry-picture"
            class="img-thumbnail"
        />
        <img
            v-if="localImage"
            :src="localImage"
            alt="entry-picture"
            class="img-thumbnail"
        />
    </template>
</template>

<script>
import { defineAsyncComponent } from "vue";
import { mapGetters, mapActions } from 'vuex';
import Swal from 'sweetalert2';

import getDayMonthYear from '@/modules/daybook/helpers/getDayMonthYear';
import uploadImage from '@/modules/daybook/helpers/uploadImage';

export default {
    name: 'EntryView',
    props: {
        id: {
            type: String,
            required: true
        }
    },
    components: {
        Fab: defineAsyncComponent(() => import('@/modules/daybook/components/Fab.vue'))
    },
    data() {
        return {
            entry: null,
            localImage: null,
            file: null
        }
    },
    computed: {
        ...mapGetters('journal', ['getEntryById']),
        day() {
            const { day } = getDayMonthYear(this.entry.date);
            return day;
        },
        month() {
            const { month } = getDayMonthYear(this.entry.date);
            return month;
        },
        yearDay() {
            const { yearDay } = getDayMonthYear(this.entry.date);
            return yearDay;
        }
    },
    methods: {
        ...mapActions('journal', ['updateEntry', 'createEntry', 'deleteEntry']),
        loadEntry(){
            let entry;
            if (this.id === 'new') {
                entry = {
                    text: '',
                    date: new Date().getTime()
                }
            } else {
                entry = this.getEntryById(this.id);
                if (!entry) return this.$router.push({name: 'no-entry'})
            }
            this.entry = entry;
        },
        async saveEntry(){
            if (!this.entry.text) {
                alert('You must write something before save it');
                return;
            }
            Swal.fire({
                title: 'Please wait...',
                allowOutsideClick: false
            });
            Swal.showLoading();
            const pictureUrl  = await uploadImage(this.file);
            if (pictureUrl) {
                this.entry.picture = pictureUrl;
                this.localImage = pictureUrl;
            }
            if (this.entry.id) { // If Id does exist then update it
                await this.updateEntry(this.entry);
            } else { // If not, create it
                const newEntryId = await this.createEntry(this.entry);
                this.$router.push({ name: 'entry', params: { id: newEntryId }})
            }
            this.file = null;
            this.localImage = null;
            Swal.fire('Saved!', 'Entry saved successfully', 'success');
        },
        async onDeleteEntry() {
            const { isConfirmed } = await Swal.fire({
                title: 'Are you sure?',
                text: 'Once deleted can not be retrieved',
                showDenyButton: true,
                confirmButtonText: 'Yes, I am sure'
            });
            if (isConfirmed) {
                Swal.fire({
                    title: 'Please wait...',
                    allowOutsideClick: false
                });
                await this.deleteEntry(this.entry.id);
                Swal.fire('Deleted!', 'Entry deleted successfully', 'success');
                this.$router.push({ name: 'no-entry' });
            }
        },
        onSelectedImage(event) {
            // console.log('onSelectedImage:event:', event.target.files[0]);
            const file = event.target.files[0];
            if (!file) {
                this.localImage = null;
                this.file = null;
                return;
            }
            this.file = file;
            const fileReader = new FileReader();
            fileReader.onload = () => this.localImage = fileReader.result;
            fileReader.readAsDataURL(file);
        },
        onSelectImage() {
            this.$refs.imageSelector.click();
        }
    },
    created() {
        this.loadEntry();
    },
    watch: {
        id() {
            this.loadEntry();
        }
    }
}
</script>

<style lang="scss" scoped>
    // .entry-title {}
    textarea {
        font-size: 20px;
        border: none;
        height: 100%;
        resize: none;
        &:focus {
            outline: none;
        }
    }
    img {
        width: 200px;
        position: fixed;
        bottom: 150px;
        right: 20px;
        box-shadow: 0px 5px 10px rgba($color: #000000, $alpha: 0.2);
    }
</style>
