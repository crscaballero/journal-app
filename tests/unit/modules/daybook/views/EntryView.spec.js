import { createStore } from "vuex";
import { shallowMount } from "@vue/test-utils";
import Swal from "sweetalert2";

import EntryView from "@/modules/daybook/views/EntryView.vue";
import journal from '@/modules/daybook/store/journal';
import { journalState } from "../../../mock-data/test-journal-state";


const createVuexStore = ( initialState ) => createStore({
    modules: {
        journal: {
            ...journal,
            state: { ...initialState }
        }
    }
});

jest.mock('sweetalert2', () => ({
    fire: jest.fn(),
    showLoading: jest.fn(),
    close: jest.fn()
}));

describe('EntryView.vue', () => {
    const store = createVuexStore(journalState);
    store.dispatch = jest.fn();
    const mockRouter = {
        push: jest.fn()
    };
    let wrapper;
    beforeEach(() => {
        jest.clearAllMocks()
        wrapper = shallowMount( EntryView, {
            props: {
                id: journalState.entries[0].id
            },
            global: {
                mocks: {
                    $router: mockRouter
                },
                plugins: [ store ]
            }
        })
    })
    test('It must redirect the user because the id doesn\'t exist', () => {
        const wrapper = shallowMount(EntryView, {
            props: {
                id: 'ThisIdDoesNotExistInTheStore'
            },
            global: {
                mocks: {
                    $router: mockRouter
                },
                plugins: [store]
            }
        });
        expect(wrapper.html()).toMatchSnapshot();
        expect(mockRouter.push).toHaveBeenCalledWith({name: 'no-entry'});
    });
    test('Must show the entry correctly', () => {
        expect(wrapper.html()).toMatchSnapshot();
        expect(mockRouter.push).not.toHaveBeenCalled();
    });
    test('It must delete the entry and get out', (done) => {
        Swal.fire.mockReturnValueOnce(Promise.resolve({ isConfirmed: true }));
        wrapper.find('.btn-danger').trigger('click');
        expect(Swal.fire).toHaveBeenCalledWith({
            title: 'Are you sure?',
            text: 'Once deleted can not be retrieved',
            showDenyButton: true,
            confirmButtonText: 'Yes, I am sure'
        });
        setTimeout(() => {
            expect(store.dispatch).toHaveBeenCalledWith('journal/deleteEntry', journalState.entries[0].id);
            expect(mockRouter.push).toHaveBeenCalled();
            done();
        }, 1);
    });
});
