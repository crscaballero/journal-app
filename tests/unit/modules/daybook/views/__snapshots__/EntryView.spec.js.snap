// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EntryView.vue It must redirect the user because the id doesn't exist 1`] = `<!--v-if-->`;

exports[`EntryView.vue Must show the entry correctly 1`] = `
<div class="entry-title d-flex justify-content-between p-2">
  <div><span class="text-success fs-3 fw-bold">4</span><span class="mx-1 fs-3">August</span><span class="textmx-2 fs-4 fw-light">2022, Thursday</span></div>
  <div><button type="button" class="btn btn-danger mx-2"> Delete <i class="fa fa-trash-all"></i></button><input type="file" accept="image/png, image/jpeg" style="display: none;"><button type="button" class="btn btn-primary"> Upload image <i class="fa fa-upload"></i></button></div>
</div>
<hr>
<div class="d-flex flex-column px-3 h-75"><textarea cols="30" rows="10" placeholder="What happened today?..."></textarea></div>
<fab-stub icon="fa-save"></fab-stub>
<img src="https://res.cloudinary.com/crscaballero/image/upload/v1659671690/journal_vue/sgfoa5nr6kyobpppahlp.jpg" alt="entry-picture" class="img-thumbnail">
<!--v-if-->
`;
